### ISSUE TITLE
`@polkadot/*` packages should be published in topological order

### ISSUE THREAD

**Issue by josepot** [polkadot-js/dev#1139]
As a result of polkadot-js/api#5841 we learned that the CI that publishes a new release for `@polkadot/api`, doesn't publish the packages in topological order. Instead, it seems to first publish the `@polkadot/api` package and then it starts publishing its dependencies.

For instance, version `@polkadot/api-augment@10.12.4` was published at `19/03/204, 14:22:35`:

![image](https://github.com/polkadot-js/api/assets/8620144/5426db7a-8d4c-485b-898e-0b54d6eef83b)

and version `@polkadot/api@10.12.4` was published 21 seconds earlier:

![image](https://github.com/polkadot-js/api/assets/8620144/1ba6abb7-104b-4494-bcff-70d97cec203b)

That means that for a span of 21 seconds the package `@polkadot/api` was "broken". This is obviously not a big deal when the release works, because if you are so unlucky to install the package during that short span of time, once you retry then everything works... However, if there is an issue during the release process (like what it happened on polkadot-js/api#5841), then the release stays broken for hours, and that is a lot more annoying to deal with.

On the other hand, if `@polkadot/api-augment` had been published before, and then the release halts for some reason, then that can't hurt anyone.

**Comment by TarikGul** [comment]
Josep and I have discussed the following, but I wanted to note some of the specifics in this issue for transparency. 

I started to write a topological sort for the `@polkadot/*` packages, and found that circular dependencies are blocking any accurate sorting.

The following logs just show the edges for the `@polkadot/api` packages. Which has multiple issues in terms of ciruclar edges.

```
[
  [ 'api', 'api-augment' ],
  [ 'api', 'api-base' ],
  [ 'api', 'api-derive' ],
  [ 'api', 'rpc-augment' ],
  [ 'api', 'rpc-core' ],
  [ 'api', 'rpc-provider' ],
  [ 'api', 'types' ],
  [ 'api', 'types-augment' ],
  [ 'api', 'types-codec' ],
  [ 'api', 'types-create' ],
  [ 'api', 'types-known' ],
  [ 'api-augment', 'api-base' ],
  [ 'api-augment', 'rpc-augment' ],
  [ 'api-augment', 'types' ],
  [ 'api-augment', 'types-augment' ],
  [ 'api-augment', 'types-codec' ],
  [ 'api-base', 'rpc-core' ],
  [ 'api-base', 'types' ],
  [ 'api-contract', 'api' ],
  [ 'api-contract', 'api-augment' ],
  [ 'api-contract', 'types' ],
  [ 'api-contract', 'types-codec' ],
  [ 'api-contract', 'types-create' ],
  [ 'api-derive', 'api' ],
  [ 'api-derive', 'api-augment' ],
  [ 'api-derive', 'api-base' ],
  [ 'api-derive', 'rpc-core' ],
  [ 'api-derive', 'types' ],
  [ 'api-derive', 'types-codec' ],
  [ 'rpc-augment', 'rpc-core' ],
  [ 'rpc-augment', 'types' ],
  [ 'rpc-augment', 'types-codec' ],
  [ 'rpc-core', 'rpc-augment' ],
  [ 'rpc-core', 'rpc-provider' ],
  [ 'rpc-core', 'types' ],
  [ 'rpc-provider', 'types' ],
  [ 'rpc-provider', 'types-support' ],
  [ 'typegen', 'api' ],
  [ 'typegen', 'api-augment' ],
  [ 'typegen', 'rpc-augment' ],
  [ 'typegen', 'rpc-provider' ],
  [ 'typegen', 'types' ],
  [ 'typegen', 'types-augment' ],
  [ 'typegen', 'types-codec' ],
  [ 'typegen', 'types-create' ],
  [ 'typegen', 'types-support' ],
  [ 'types', 'types-augment' ],
  [ 'types', 'types-codec' ],
  [ 'types', 'types-create' ],
  [ 'types-augment', 'types' ],
  [ 'types-augment', 'types-codec' ],
  [ 'types-create', 'types-codec' ],
  [ 'types-known', 'types' ],
  [ 'types-known', 'types-codec' ],
  [ 'types-known', 'types-create' ]
]
```

Next if you are to run `yarn polkadot-dev-circular` in `@polkadot/api` repo, you will notice that the following repository catches: `FATAL: Failed with 425 circular dependencies`.

This becomes a big issue for creating a safer release process. Currently my idea is to still topologically sort, but instead of the sorting algorithm breaking at anything circular deps it takes an approach of "safer" ordering.

**Comment by TarikGul** [comment]
Released in 0.79.1
